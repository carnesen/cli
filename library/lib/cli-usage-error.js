"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliUsageError = exports.CLI_USAGE_ERROR = void 0;
exports.CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';
/**
 * A custom error with code {@linkcode CLI_USAGE_ERROR}
 * @remarks
 * {@linkcode runCliAndExit} catches this class of error and prints command-line usage
 * (i.e. "Usage: cloud users lists ...") for the current command. If the `message`
 * property is provided, `runCliAndExit` will normal `Error`. See also print `Error: <your
 * message>`. See also {@linkcode CliTerseError}.
 * @example
 * ```typescript
 * throw new CliUsageError("Expected one or more values")
 * ```
 */
class CliUsageError extends Error {
    /**
     *
     * @param message An `Error` message
     * @param commandStack Used internally for argument processing
     */
    constructor(message, commandStack) {
        super(message);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandStack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.code = exports.CLI_USAGE_ERROR;
        this.commandStack = commandStack;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CliUsageError = CliUsageError;
//# sourceMappingURL=cli-usage-error.js.map