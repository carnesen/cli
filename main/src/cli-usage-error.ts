import { Node } from './cli-node';

/** "code" of a {@link CliUsageError} */
export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

/**
 * Thrown to print command usage and an error message but not a `stack`
 */
export class CliUsageError extends Error {
  /** The string constant {@linkcode CLI_USAGE_ERROR} */
  public readonly code: typeof CLI_USAGE_ERROR;

  /** Used internally for constructing the command-line usage string */
  public node?: Node;

  /**
   * @param message If provided, {@linkcode runCliAndExit} will also print "Error: \<your
   * message\>"
   * @param node Used internally for constructing the command-line usage string
   */
  constructor(message?: string, node?: Node) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.node = node;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
