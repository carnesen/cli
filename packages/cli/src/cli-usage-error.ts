import { ICliTree } from './cli-tree';

/** "code" of a {@link CliUsageError} */
export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

/**
 * Thrown to print usage documentation and an error message
 */
export class CliUsageError extends Error {
	/** The string constant [[`CLI_USAGE_ERROR`]] */
	public readonly code: typeof CLI_USAGE_ERROR;

	/** Used internally for generating usage documentation */
	public tree?: ICliTree;

	/**
	 * @param message If provided, [[`ICli.run`]] will also print "Error: \<your
	 * message\>"
	 * @param tree Used internally for constructing the command-line usage string
	 */
	public constructor(message?: string, tree?: ICliTree) {
		super(message);
		this.code = CLI_USAGE_ERROR;
		this.tree = tree;
		Object.setPrototypeOf(this, new.target.prototype);
	}
}
