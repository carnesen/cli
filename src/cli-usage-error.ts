import { CommandStack } from './types';

export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

export class CliUsageError extends Error {
  public readonly code: typeof CLI_USAGE_ERROR;

  public commandStack?: CommandStack;

  constructor(message?: string, commandStack?: CommandStack) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.commandStack = commandStack;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
