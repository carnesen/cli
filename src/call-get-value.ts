import { AnyArgParser } from './types';
import { CliUsageError } from './cli-usage-error';

export async function callGetValue(
  argParser: AnyArgParser,
  args?: string[],
  context?: string,
): Promise<any> {
  const { required, placeholder, getValue } = argParser;
  let prefix = [context, placeholder].filter(Boolean).join(' ');
  if (prefix) {
    prefix += ' : ';
  }
  if (required && (!args || args.length === 0)) {
    throw new CliUsageError(`${prefix}ArgParser is required`);
  }
  try {
    return await getValue(args);
  } catch (exception) {
    if (exception && typeof exception.message === 'string') {
      exception.message = `${prefix}${exception.message}`;
    }
    throw exception;
  }
}
