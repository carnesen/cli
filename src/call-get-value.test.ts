import { runAndCatch } from '@carnesen/run-and-catch';

import { callGetValue } from './call-get-value';
import { CLI_USAGE_ERROR } from './cli-usage-error';
import {
  dummyArgParser,
  dummyRequiredArgParser,
  DUMMY_INPUT_THROWN_INTENTIONALLY,
  DUMMY_INPUT_THROW,
  DUMMY_INPUT_THROW_NON_TRUTHY,
} from './dummy-inputs-for-testing';

describe(callGetValue.name, () => {
  it(`returns getValue(argv) if an argv with length >= 1 is passed`, async () => {
    const argv = ['foo'];
    expect(await callGetValue(dummyArgParser, argv)).toBe(dummyArgParser.getValue(argv));
    expect(await callGetValue(dummyRequiredArgParser, argv)).toBe(
      dummyRequiredArgParser.getValue(argv),
    );
  });

  it(`if not required, returns getValue(argv) if argv is an empty array or undefined`, async () => {
    expect(await callGetValue(dummyArgParser, [])).toBe(dummyArgParser.getValue([]));
    expect(await callGetValue(dummyArgParser, undefined)).toBe(
      dummyArgParser.getValue(undefined),
    );
  });

  it(`if required, throws usage error "argParser is required" if argv is an empty array or undefined`, async () => {
    for (const argv of [undefined, [] as string[]]) {
      const exception = await runAndCatch(callGetValue, dummyRequiredArgParser, argv);
      expect(exception.code).toBe(CLI_USAGE_ERROR);
      expect(exception.message).toMatch(/ArgParser is required/i);
      expect(exception.message).toMatch(dummyRequiredArgParser.placeholder);
    }
  });

  it(`if throws "argParser is required", expect message to match snapshot`, async () => {
    const exception = await runAndCatch(callGetValue, dummyRequiredArgParser);
    expect(exception.message).toMatchSnapshot();
  });

  it(`if throws "argParser is required" with context, expect message to match snapshot`, async () => {
    const exception = await runAndCatch(
      callGetValue,
      dummyRequiredArgParser,
      undefined,
      'context',
    );
    expect(exception.message).toMatchSnapshot();
  });

  it(`throws if getValue does with a context/placeholder enhanced message`, async () => {
    const exception = await runAndCatch(callGetValue, dummyArgParser, [
      DUMMY_INPUT_THROW,
    ]);
    expect(exception.message).toMatch(DUMMY_INPUT_THROWN_INTENTIONALLY);
    expect(exception.message).toMatch(dummyArgParser.placeholder);
    expect(exception.message).toMatchSnapshot();
  });

  it(`just re-throws exception if getValue throws a non-truthy exception`, async () => {
    const exception = await runAndCatch(callGetValue, dummyArgParser, [
      DUMMY_INPUT_THROW_NON_TRUTHY,
    ]);
    expect(exception).not.toBeTruthy();
  });
});
