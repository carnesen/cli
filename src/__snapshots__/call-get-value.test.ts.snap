// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callGetValue if throws "argParser is required" with context, expect message to match snapshot 1`] = `"context <foo> : ArgParser is required"`;

exports[`callGetValue if throws "argParser is required", expect message to match snapshot 1`] = `"<foo> : ArgParser is required"`;

exports[`callGetValue throws if getValue does with a context/placeholder enhanced message 1`] = `"thrown intentionally"`;
